presto-hive/src/test/java/com/facebook/presto/hive/statistics/TestQuickStatsProvider.java

```diff
@@ -0,0 +1,342 @@
+                new SchemaTableName(TEST_SCHEMA, TEST_TABLE), metastoreContext, testPartitions), commonPool());
+
+        CompletableFuture<Map<String, PartitionStatistics>> future2 = supplyAsync(() -> quickStatsProvider.getQuickStats(SESSION, metastoreMock,
+                new SchemaTableName(TEST_SCHEMA, TEST_TABLE), metastoreContext, testPartitions), commonPool());
+
+        allOf(future1, future2).join();
+
+        // Verify only one call was made for each test partition
+        for (String testPartition : testPartitions) {
+            verify(longRunningQuickStatsBuilderMock).buildQuickStats(any(), any(), any(), any(), eq(testPartition), any());
+        }
+        verifyNoMoreInteractions(longRunningQuickStatsBuilderMock);
+
+        Map<String, PartitionStatistics> quickStats1 = future1.get();
+        Map<String, PartitionStatistics> quickStats2 = future2.get();
+
+        // Both PartitionStatistics have stats for all partitions
+        assertEquals(quickStats1.entrySet().size(), testPartitions.size());
+        assertTrue(quickStats1.keySet().containsAll(testPartitions));
+        assertEquals(quickStats2.entrySet().size(), testPartitions.size());
+        assertTrue(quickStats2.keySet().containsAll(testPartitions));
+
+        // For the same partition, we will observe that one of them has the expected partition stats and the other one is empty
+        for (String testPartition : testPartitions) {
+            PartitionStatistics partitionStatistics1 = quickStats1.get(testPartition);
+            PartitionStatistics partitionStatistics2 = quickStats2.get(testPartition);
+
+            if (partitionStatistics1.equals(empty())) {
+                assertEquals(partitionStatistics2, expectedPartitionStats);
+            }
+            else if (partitionStatistics2.equals(empty())) {
+                assertEquals(partitionStatistics1, expectedPartitionStats);
+            }
+            else {
+                fail(String.format("For [%s] partitionExpected one of the partitions stats to be empty. Actual partitionStatistics1 [%s], partitionStatistics2 [%s]",
+                        testPartition, partitionStatistics1, partitionStatistics2));
+            }
+        }
+
+        // Future calls for the same partitions will return from cached partition stats with valid values
+        Map<String, PartitionStatistics> quickStats = quickStatsProvider.getQuickStats(SESSION, metastoreMock,
+                new SchemaTableName(TEST_SCHEMA, TEST_TABLE), metastoreContext, testPartitions);
+
+        // Verify only one call was made for each test partition
+        assertEquals(quickStats.entrySet().size(), testPartitions.size());
+        assertTrue(quickStats.keySet().containsAll(testPartitions));
+        quickStats.values().forEach(ps -> assertEquals(ps, expectedPartitionStats));
+
+        verifyNoMoreInteractions(longRunningQuickStatsBuilderMock);
+    }
+
+    @Test
+    public void quickStatsBuildTimeIsBounded()
```

Comment:
Do we need a test for cache expiry as well?

Reply:
Since we're using a Guava Cache with well tested expiry behavior, I chose to skip this test